### DISCUSSION TITLE
회원 및 비회원 로직

### DISCUSSION THREAD

**Discussion by woo-chang** [woowacourse-teams/2023-votogether#256] [자유로운 토론]
# 주제

게시글 목록 조회, 댓글 목록 조회 등과 같은 기능은 사용자 유입을 위해 회원가입 하지 않아도 제공하도록 기획하였습니다.
회원과 비회원을 어떻게 구분하고 각각의 기능은 어떻게 구현하면 좋을지 생각을 정리해보면 좋을 것 같아요 :)

**Comment by jeomxon** [comment]
토큰 등으로 하나의 API에서 분기처리를 하기는 제 머리로는 도저히 떠오르지 않네요...
따라서 API를 분리하여 따로 호출하면 좋을 것 같아요.
아예 같은 것도 있을거고 비슷한 로직도 있겠지만, 내부적으로는 함수를 재활용할 수 있기 때문에 크게 문제없을 것 같아요.
그러면 인가에 대한 문제로부터도 자유로워지지 않을까라는 생각이 듭니다.

**Comment by woo-chang** [comment]
제 개인적인 생각으로는 비회원, 회원 API 분리하는게 좋을 것 같아요! 그 이유는 아래와 같습니다.
보통 경로로 **인증이 필요한 API**와 **인증이 필요하지 않은 API**를 구분하는데 인증이 필요한 경로임에도 인증없이 사용할 수 있다면 사용하는 입장에서 이해하기 어려울 수 있다고 생각해요. 또한 내부적으로는 조회라는 비슷한 기능이지만 두 개의 API의 역할이 다르다고 생각합니다.
추가적으로 API는 최대한 한 가지의 일만 하도록 하여 분기가 없는 것이 나중에 유지보수하기에도 용이할 것이라 생각해요 :)

**Comment by tjdtls690** [comment]
당장 생각나는 방법은,

현재 만들어놓은 조회 메서드들을 재활용할 때, 투표 관련 데이터들은 전부 -1을 반환하게 하는 비회원용 API를 나누는 것이네요. 그리고 비회원용 API URL은 필터에서 안걸러지게끔 설정해놓으면 될 듯 합니다.

그런데 다른 기능들을 클릭할 땐 비회원이란 것을 어떻게 구분지어줄 수 있을까요? 비회원 상태에서 글 작성 버튼같은 걸 누르면 필터에서 예외를 던지게 될 것 같은데, 프론트에서 애초부터 버튼을 없애버리는 건가요?

**Reply by woo-chang** [reply]
백엔드, 프론트 모두 검증이 필요할 것 같아요 :)

프론트에서는 로그인이 되어있지 않을 때(인증 토큰이 존재하지 않을 때) 버튼이 보이지 않도록 하는 작업이 필요할 것 같아요.
백엔드에서도 API를 버튼을 눌렀을 때만 호출되는 것이 아니라 해당 API 엔드포인트를 통해 언제든 HTTP 요청이 가능하기에, 비회원일 때(인증 토큰이 없을 때) 인증 토큰이 필요한 API를 호출한다면 예외가 발생하도록 코드를 작성해야 할 것 같아요.

쌍방 검증을 통해 안정성있는 애플리케이션을 만들 수 있을 것 같습니다 😀

**Comment by aiaiaiai1** [comment]
저는 비회원, 회원API 를 분리하는게 더 좋은 거 같아요
이 두개의 api를 분리함으로서 명확하게 비회원과 회원의 기능을 구별하여 확인할수 있고
카테고리 목록을 조회하는 기능에서도 비회원과 회원으로 나누어서 api를 만들었기 때문에 그부분에 대해서도 일관성을 통일할수 있을것 같다는 생각입니다.
